Partition spec parser improvements

bashtanov opened the pull request:
Improve to support the grammar in full.

Syntax: https://iceberg.apache.org/docs/nightly/spark-ddl/#partitioned-by (top-level brackets are optional)
Supported transformations: https://iceberg.apache.org/spec/#partitioning

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
iceberg/transform: make transform types know their names as strings

both for parsing and for partition name generation

bashtanov committed:
utils/fixed_string: implement concatenation

bashtanov committed:
utils/variant: compile-time type manipulation

For a variant type V implement two utility types:
1) variant_of_identities<V>: a variant of std::type_identity<T> for each
type T in V. Use-case: store variant choice only without variant data,
neater than numeric index.
2) tuple_of_identities<V>: a tuple of std::type_identity<T> for each
type T in V. Use-case: generate compile-time loop over all options of V
with folding operator.

bashtanov committed:
iceberg/unresolved_partition_spec: function to generate partition name

If partition name is not specified in the spec we need to generate it.
Implement in accordance to Iceberg reference implementation.

bashtanov committed:
datalake/partition_spec_parser: implement the grammar in full

bashtanov committed:
datalake/tests/partition_spec_parser: cover more cases

bashtanov commented:
/dt

bashtanov commented:
/dt

bashtanov commented:
/dt

bashtanov commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62531-->
<!--646b4407-8fbc-4191-a77b-9b1c4c4c2052-->
<!--30f7fb6b-a1e0-4d69-b827-7853ff399bbd-->
<!--78281ed0-ab91-4062-b791-61059ea1e15f-->
<!--f857df28-6ef2-4755-8e1b-755b1ff4f8ac-->
<!--title-->
## Retry command for [Build#62531](https://buildkite.com/redpanda/redpanda/builds/62531)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/datalake/custom_partitioning_test.py::DatalakeCustomPartitioningConfigTest.test_configs
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_illegal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"spark","test_case":"drop column that appears in partition spec"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"spark","test_case":"add_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"spark","test_case":"drop_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"spark","test_case":"promote_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_illegal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"trino","test_case":"drop column that appears in partition spec"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"trino","test_case":"add_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"trino","test_case":"drop_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"trino","test_case":"promote_column"}
tests/rptest/tests/datalake/custom_partitioning_test.py::DatalakeCustomPartitioningTest.test_basic@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_illegal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"spark","test_case":"drop column that appears in partition spec"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"spark","test_case":"add_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"spark","test_case":"promote_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"spark","test_case":"reorder_columns"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_illegal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"trino","test_case":"drop column that appears in partition spec"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"trino","test_case":"add_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"trino","test_case":"promote_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"trino","test_case":"reorder_columns"}
tests/rptest/tests/datalake/custom_partitioning_test.py::DatalakeCustomPartitioningTest.test_basic@{"catalog_type":"rest_hadoop","cloud_storage_type":1}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_illegal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"spark","test_case":"drop column that appears in partition spec"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"spark","test_case":"add_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"spark","test_case":"drop_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"spark","test_case":"reorder_columns"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"trino","test_case":"add_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"trino","test_case":"drop_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"trino","test_case":"reorder_columns"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"spark","test_case":"drop_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"spark","test_case":"promote_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"spark","test_case":"reorder_columns"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_illegal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"trino","test_case":"drop column that appears in partition spec"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"avro","query_engine":"trino","test_case":"drop_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto2","query_engine":"trino","test_case":"promote_column"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_legal_schema_evolution@{"cloud_storage_type":1,"produce_mode":"proto3","query_engine":"trino","test_case":"reorder_columns"}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"trino","use_partition_spec":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3d6f4d8f-be81-4d08-af14-70def6e40b6b-->
<!--TEST RESULTS-->
<!--a3a836ea-954b-47d5-998e-8cd8ae36a3d2-->
<!--TEST RESULTS-->
<!--85e8dac3-9fe9-4b24-bc09-07e939d7304e-->
<!--TEST RESULTS-->
<!--c9c92e3d-1883-4873-9eae-c3d46a8954b7-->
<!--TEST RESULTS-->
<!--1199364d-910b-4940-b2c1-01007ca44aeb-->
<!--TEST RESULTS-->
<!--2beb4b6d-5f8d-4749-bbe8-8ebc2f1969eb-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62531">build#62531</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| iceberg_compatibility_rpunit.iceberg_compatibility_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62531#01955b41-6fde-4495-adc2-1d7287d8609b | FAIL | 0/2 |
| partition_spec_parser_rpunit.partition_spec_parser_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62531#01955b41-6fde-4495-adc2-1d7287d8609b | FAIL | 0/2 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningConfigTest.test_configs | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=dropcolumnthatappearsinpartitionspec.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=dropcolumnthatappearsinpartitionspec.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=dropcolumnthatappearsinpartitionspec.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=dropcolumnthatappearsinpartitionspec.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=dropcolumnthatappearsinpartitionspec.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_illegal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=dropcolumnthatappearsinpartitionspec.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=add_column.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=add_column.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=add_column.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=drop_column.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=drop_column.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=drop_column.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=promote_column.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=reorder_columns.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=reorder_columns.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.test_case=reorder_columns.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=add_column.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=add_column.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=add_column.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=drop_column.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=drop_column.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=drop_column.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=promote_column.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=promote_column.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=promote_column.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-4036-896d-d03213ae71a4 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=reorder_columns.produce_mode=ProducerType.AVRO | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=reorder_columns.produce_mode=ProducerType.PROTO2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c43f-46b1-b715-ccf7e6a39d2e | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_legal_schema_evolution.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.test_case=reorder_columns.produce_mode=ProducerType.PROTO3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c441-449c-bc63-c6d0c704d24f | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FAIL | 0/20 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=True.server_cfg=None.topic_cfg=300000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955b9d-c440-42ad-a580-67d9f121f8cf | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955ba1-1caf-4e2d-a79a-409d2d9cce73 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62531#01955ba1-1caf-4e2d-a79a-409d2d9cce73 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62556">build#62556</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62556#01955d55-61b7-4cca-b856-b9f7d042697b | FLAKY | 1/2 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62556#01955d55-61b7-4e64-a369-804b6deac92d | FLAKY | 1/2 |
| rptest.tests.rpk_registry_test.RpkRegistryTest.test_produce_consume_avro | ducktape | https://buildkite.com/redpanda/redpanda/builds/62556#01955d55-61b7-4cca-b856-b9f7d042697b | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/62556#01955d55-61b6-4fc4-a427-2b029a4c505b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62857">build#62857</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62857#019580e8-d93d-4b02-9a06-d68fb9e5ad8c | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62857#019580e8-d93d-4b02-9a06-d68fb9e5ad8c | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62857#01958104-4606-4dc0-933b-e6cd9ec3d1c4 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62902">build#62902</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62902#0195846a-e51d-4348-a5f2-83db9d652acc | FLAKY | 1/2 |
| rptest.tests.archive_retention_test.CloudArchiveRetentionTest.test_delete.cloud_storage_type=CloudStorageType.ABS.retention_type=retention.ms | ducktape | https://buildkite.com/redpanda/redpanda/builds/62902#019584c9-5eb8-4b02-98fe-a6450fbc146b | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62902#019584c4-d385-4b02-8f8f-c98f0e121288 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62902#019584c4-d385-4530-ade4-f9ff1036bd0a | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62902#019584c4-d384-4a0c-96f4-a4eb8a49b62c | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62902#019584c4-d386-4cd0-b73d-50ec246cc4ac | FLAKY | 1/4 |
| translator_test_rpfixture.translator_test_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62902#0195846a-e51c-46ba-8182-9b97f3552974 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62912">build#62912</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585b5-e496-49ae-97fe-55d70ac716a3 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585b5-e497-4d82-a7c1-96bd8b730dd3 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585b5-e495-4eaa-9609-0ef4181d00ac | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585b5-e496-49ae-97fe-55d70ac716a3 | FLAKY | 1/3 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585d2-25ee-415a-a55f-b9564d05a06a | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585d2-25ee-415a-a55f-b9564d05a06a | FLAKY | 1/2 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=None.allow_gaps_cluster_level=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62912#019585d2-25f0-4f89-9923-6298174c3c10 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62978">build#62978</a></summary>


| test_id                                                                                                                                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62978#01958a89-79d5-482f-b03b-a824ed86329d | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62978#01958a85-4cc9-4bbf-9d43-82c137930dcc | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62978#01958a85-4cc9-4bbf-9d43-82c137930dcc | FLAKY         | 1/2      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62978#01958a89-79d5-482f-b03b-a824ed86329d | FLAKY         | 1/2      |
</details>
<!--/body-->

bashtanov committed:
utils/fixed_string: support 0-length string, remove `const char*` cast

Cast to `const char*` was returning a pointer to an array without
terminating zero byte. This conversion is not in use and looks unsafe,
remove.

While certain compilers support them as extensions, 0-size arrays are
disallowed in C++. Clang is okay but ASAN is not happy about one.
Implement `fixed_string<0>` separately.

bashtanov commented:
/dt

bashtanov commented:
/dt

ztlpn commented on src/v/iceberg/transform.h:
```diff
@@ -9,24 +9,34 @@
+struct identity_transform : public transform_base<"identity"> {
+    static constexpr fixed_string suffix{""};
+};
+struct bucket_transform : public transform_base<"bucket"> {
```
Why do we need inheritance here? Can't this just be a simple visitor?

bashtanov replied:
Do you mean storing these string literals in some function, not in the types themselves?

bashtanov replied:
removed inheritance as discussed in the meeting

ztlpn commented on src/v/datalake/tests/partition_spec_parser_test.cc:
```diff
@@ -12,64 +12,157 @@
+} // namespace
+
+TEST(PartitionSpecParserTest, TestParse) {
+    expect_success({}, {"", "()", "   (  )  ", "\t\r\n"});
```
Hmm I thought we converged on requiring the parentheses?

bashtanov replied:
We converged on no special meaning for empty string vs empty parentheses. This renders outermost parentheses meaningless, so I don't see a point in requiring them. Do you?

ztlpn replied:
Chatted offline and agreed to make parentheses optional.

ztlpn commented on src/v/datalake/tests/partition_spec_parser_test.cc:
```diff
@@ -12,64 +12,157 @@
-          unresolved_partition_spec{.fields = std::move(expected)});
-    }
+void expect_failure(
+  std::string_view marked_input,
```
Nit: doc comment for this function explaining the parameter syntax and meaning would be nice.

ztlpn commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
+          : res(std::in_place, std::move(val), next) {}
+
+        // success
+        parse_result(
```
What is the meaning of this constructor? Why do we log expectation and error position if it is a success?

bashtanov replied:
That's to indicate we have extracted a value of required type, but failed to extract a longer one of the same type, and it happened at _this_ position by failing to find _this_ token. Then we attempt other paths, and if they all break no further than the same position we should include this info in the diagnostics.

E.g. we have a sequence of identifiers delimited with a full stop, and after one of them we don't see the full stop. We return the sequence but still remember where we failed. Then we try other paths such as a "AS _alias_", a comma preceding next field definition, a closing outermost bracket etc. If they all fail we report that at the position we found neither a full stop to continue fully-qualified identifier sequence, nor AS/comma/etc.

ztlpn commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
+                    ss::sstring res{};
+                    position cur2 = pos;
+                    while (cur2 != cur) {
+                        res.append(&*cur2, 1);
```
can it be a simple push_back?

bashtanov replied:
is there such a method in sstring?

ztlpn replied:
looks like append() copies the whole string...

bashtanov replied:
changed to pre-allocate required length

ztlpn commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
+        while (cur != end()) {
+            if (*cur == quotation_mark) {
+                ++cur;
+                auto dbl_q_res = expect_with_ws<quotation_mark_as_array>(cur);
```
Why do we allow whitespace here? IIUC,
```
`foo`   `bar`
```
will get parsed as 
```
foo`bar
```
, right? It would make more sense to me to require backticks to be immediately next to each other for escaping.

bashtanov replied:
thanks, I'll fix that and will add a test on this case

ztlpn commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
-    skip_space(unparsed);
+    parse_result<iceberg::unresolved_partition_spec::field>
+    parse_transform_expr(position pos) {
+        auto transofrm_name_res = parse_transform_name(pos);
```
typo

dotnwat commented on src/v/utils/variant.h:
```diff
@@ -0,0 +1,37 @@
+
+#include <variant>
+
+namespace {
```
what's the purpose of an anonymous namespace in a header?

bashtanov replied:
I separated it into a reusable header only as a very last stage and didn't think about it. I'll change it, thank you.

dotnwat commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
+          parser_base& self,
+          expectation exp,
+          position err_pos)
+          : res(std::in_place, std::move(val), next) {
```
did you intend to forward from val?

bashtanov replied:
I did not intend to instantiate `parse_result` or `parse_success` templates with a reference-type `T` TBH. It may make sense in some weird situation, e.g. where the result of parsing is a reference to something present before parsing and outliving it. So I'll change it to forward.

dotnwat commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
-    auto it = str.begin();
-    while (it != str.end() && std::isspace(*it)) {
-        ++it;
+        T& value_lref() & {
```
`mutable_value()`?

dotnwat commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,516 @@
+    parse_column_reference(position pos) {
+        return parse_separated_list<ss::sstring, std::vector, ".">(
+          &self::parse_identifier, pos);
+    }
 
-    return parse_result<transform_field>{
-      .val = std::move(result),
-      .unparsed = unparsed,
+    parse_result<iceberg::unresolved_partition_spec::field>
+    parse_column_reference_expr(position pos) {
+        auto res = parse_column_reference(pos);
+        if (!res) {
+            return res;
+        }
+        pos = res.after();
+        return {{.source_name = std::move(res).value()}, pos};
     };
-}
 
-std::optional<parse_result<iceberg::unresolved_partition_spec::field>>
-parse_partition_field(const std::string_view& str, parse_ctx& ctx) {
-    auto unparsed = str;
-    skip_space(unparsed);
-
-    transform_field tf;
-    if (auto parsed_tf = parse_transform_field(unparsed, ctx); parsed_tf) {
-        tf = std::move(parsed_tf->val);
-        unparsed = parsed_tf->unparsed;
-    } else if (auto parsed_sf = parse_qualified_identifier(unparsed);
-               parsed_sf) {
-        tf.source = std::move(parsed_sf->val);
-        tf.transform = iceberg::identity_transform{};
-        unparsed = parsed_sf->unparsed;
-    } else {
-        ctx.report_expected(unparsed, "qualified identifier or transform");
-        return std::nullopt;
+    using transform_ident = variant_of_identities<iceberg::transform>;
+
+    parse_result<transform_ident> parse_transform_name(position pos) {
+        auto tagged_res = parse_result<transform_ident>::dummy_failure();
```
would this commit be easier to understand if you just added all new code and removed the old code? right now all the add/removes are jumbled together.

bashtanov replied:
My change was exactly adding new code and removing old code. However, to split this commit into two separate ones I'll need to introduce some tricks to ensure it is compilable in between them.

How the change is presented is a matter of the diff engine, not an inherent property of the change. Maybe could you just click "view file" in the per-file menu in the top-right corner to view the resulting code only?

That's not unlike separating code changes from formatting changes: does it make sense if Github has an whitespace-ignoring comparison mode?

ztlpn commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,543 @@
+                    // it was a single quote: finalize!
+                    position reread_cur = pos;
+                    auto reread_end = cur - 1; // excluding the closing tick
+                    vassert(reread_cur <= reread_end, "");
```
nit: assert message

ztlpn commented on src/v/datalake/partition_spec_parser.cc:
```diff
@@ -10,260 +10,543 @@
+
+                    size_t dst_len = reread_end - reread_cur
+                                     - cnt_dbl_backticks;
+                    vassert(dst_len >= 0, "");
```
isn't the assert trivial as `dst_len` is a size_t?

bashtanov replied:
I removed both

bashtanov commented:
//src/v/storage/tests:compaction_e2e_test failed and did not save logs:
```
2025-03-11 13:54:16 UTC	FAIL: //src/v/storage/tests:compaction_e2e_test (see /root/.cache/bazel/_bazel_root/81bb5027d6e0d042eb9e7f6b3ea991f0/execroot/_main/bazel-out/k8-fastbuild/testlogs/src/v/storage/tests/compaction_e2e_test/test.log)
...
2025-03-11 13:54:16 UTC	INFO: From Testing //src/v/storage/tests:compaction_e2e_test:
2025-03-11 13:54:16 UTC	stdout (/root/.cache/bazel/_bazel_root/81bb5027d6e0d042eb9e7f6b3ea991f0/execroot/_main/bazel-out/_tmp/actions/stdout-521) 41015210 exceeds maximum size of --experimental_ui_max_stdouterr_bytes=1048576 bytes; skipping
```
https://buildkite.com/redpanda/redpanda/builds/62912#01958573-1dc5-4dda-bf93-893a326763de

bashtanov commented:
/dt
